src/monodroid/jni/jni-wrappers.h

```diff
@@ -0,0 +1,179 @@
+			  jstr (reinterpret_cast<jstring> (jo))
+		{
+			assert (env);
+		}
+
+		explicit jstring_wrapper (JNIEnv *env, const jstring js) noexcept
+			: env (env),
+			  jstr (js)
+		{
+			assert (env);
+		}
+
+		jstring_wrapper (const jstring_wrapper&) = delete;
+
+		~jstring_wrapper () noexcept
+		{
+			release ();
+		}
+
+		jstring_wrapper& operator=(const jstring_wrapper&) = delete;
+
+		const char* get_cstr () noexcept
+		{
+			if (cstr == nullptr && env != nullptr)
+				cstr = env->GetStringUTFChars (jstr, nullptr);
+
+			return cstr;
+		}
+
+		jstring_wrapper& operator= (const jobject new_jo) noexcept
+		{
+			assign (reinterpret_cast<jstring> (new_jo));
+			return *this;
+		}
+
+		jstring_wrapper& operator= (const jstring new_js) noexcept
+		{
+			assign (new_js);
+			return *this;
+		}
+
+	protected:
+		void release () noexcept
+		{
+			if (jstr == nullptr || cstr == nullptr || env == nullptr)
+				return;
+			env->ReleaseStringUTFChars (jstr, cstr);
+			jobjectRefType type = env->GetObjectRefType (jstr);
+			switch (type) {
+				case JNILocalRefType:
+					env->DeleteLocalRef (jstr);
+					break;
+
+				case JNIGlobalRefType:
+					env->DeleteGlobalRef (jstr);
+					break;
+
+				case JNIWeakGlobalRefType:
+					env->DeleteWeakGlobalRef (jstr);
+					break;
+
+				case JNIInvalidRefType: // To hush compiler warning
+					break;
+			}
+
+			jstr = nullptr;
+			cstr = nullptr;
+		}
+
+		void assign (const jstring new_js) noexcept
+		{
+			release ();
+			if (new_js == nullptr)
+				return;
+
+			jstr = new_js;
+			cstr = nullptr;
+		}
+
+		friend class jstring_array_wrapper;
+
+	private:
+		jstring_wrapper ()
+			: env (nullptr),
+			  jstr (nullptr)
+		{}
+
+	private:
+		JNIEnv *env;
+		jstring jstr;
+		const char *cstr = nullptr;
+	};
+
+	class jstring_array_wrapper
+	{
+	public:
+		explicit jstring_array_wrapper (JNIEnv *env) noexcept
+			: env (env),
+			  arr (nullptr),
+			  len (0)
+		{
+			assert (env);
+		}
+
+		explicit jstring_array_wrapper (JNIEnv *env, jobjectArray arr)
+			: env (env),
+			  arr (arr)
+		{
+			assert (env);
+			assert (arr);
+			len = env->GetArrayLength (arr);
+			if (len > sizeof (static_wrappers) / sizeof (jstring_wrapper))
+				wrappers = new jstring_wrapper [len];
+			else
+				wrappers = static_wrappers;
+		}
+
+		~jstring_array_wrapper () noexcept
+		{
+			if (wrappers != nullptr && wrappers != static_wrappers)
+				delete[] wrappers;
+		}
+
+		size_t get_length () const noexcept
+		{
+			return len;
+		}
+
+		jstring_wrapper& operator[] (size_t index) noexcept
+		{
+			if (index >= len)
+				return invalid_wrapper;
+
+			if (wrappers [index].env == nullptr) {
+				wrappers [index].env = env;
+				wrappers [index].jstr = reinterpret_cast <jstring> (env->GetObjectArrayElement (arr, index));
+			}
+
+			return wrappers [index];
+		}
+
+	private:
+		JNIEnv *env;
+		jobjectArray arr;
+		size_t len;
+		jstring_wrapper *wrappers;
+		jstring_wrapper  static_wrappers[5];
+		jstring_wrapper  invalid_wrapper;
```

Comment:
Could/should this be `static`?  I'm not sure I see much sense in "wasting" 3*sizeof(void*) bytes of space.

Reply:
They should not be `static`. That would make them shared across all the instances - dangerous and invalid (especially for the array). This is a tradeoff between memory use (not much of it used) and speed (large savings when we don't have to allocate the `wrappers` array for each instance of the class). The only one out of these three that could be static is `invalid_wrapper` but that again creates uncertainty about what happens if the returned reference is retained and used in an unknown way by some unknown future code - a static version would be shared across all the code that would get it and any modifications to it would be visible everywhere the reference is retained. In short - it introduces uncertainty for the sake of saving two `sizeof(void*)` which is not worth it IMO.

Reply:
It's for saving *three* `sizeof(void*)`, as `jstring_wrapper` has three pointer members: `env`, `jstr`, `cstr`.

And I only suggested using `static` on `invalid_wrapper`, not anything else.

Since `invalid_wrapper` will have a null `env` pointer, the instance is ~useless. (Also thread-safe, as all possibly mutating operations do nothing when `env==nullptr`.)

It's *potentially* "not worth it," but `jstring_array_wrapper` is *huge*: `3*sizeof(void*) + sizeof(size_t) + 6*3*sizeof(void*)` == *88* bytes on a 32-bit ABI, double that on a 64-bit ABI.

On the stack.

Perhaps 88 bytes isn't as much as I fear, but IIRC we only have like 2-4KB stacks on Android...

Reply:
The default stack size in bionic is 1MB, so I wouldn't worry too much...